spec/unit/network/authconfig_spec.rb

```diff
@@ -107,3 +107,29 @@
     end
   end
 end
+
+describe Puppet::Network::AuthConfig do
+  class TestAuthProvider
+    def initialize(rights=nil); end
+    def check_authorization(method, path, params); end
+  end
+
+  it "instantiates authprovider_class with rights" do
+    Puppet::Network::AuthConfig.authprovider_class = TestAuthProvider
+    rights = Puppet::Network::Rights.new
+    TestAuthProvider.expects(:new).with(rights)
+    described_class.new(rights)
+  end
+
+  it "delegates authorization check to authprovider_class" do
+    Puppet::Network::AuthConfig.authprovider_class = TestAuthProvider
+    TestAuthProvider.any_instance.expects(:check_authorization).with(:save, '/path/to/resource', {})
+    described_class.new.check_authorization(:save, '/path/to/resource', {})
+  end
+
+  it "uses DefaultAuthProvider by default" do
+    Puppet::Network::AuthConfig.authprovider_class = nil
```

Comment:
> Happy to try to reorder these tests and them change/simplify AuthConfig to have @@authprovider_class = DefaultAuthProvider.

I tried this out and it works.  There's less code.  I just moved the default test up to the top, and got rid of the setter call.  The AuthConfig would then look like:
```ruby
  class Network::AuthConfig
    @@authprovider_class = Puppet::Network::DefaultAuthProvider

    def self.authprovider_class=(klass)
      @@authprovider_class = klass
    end

    def initialize(rights=nil)
      if rights.nil?
        @authprovider = @@authprovider_class.new
      else
        @authprovider = @@authprovider_class.new(rights)
      end
    end

    def check_authorization(method, path, params)
      @authprovider.check_authorization(method, path, params)
    end
  end
```

One difference with this implementation is that it's now possible to set the authprovider_class to nil and things will blow up.  Like I said, the implementation in this PR at least protects against accidental nils by moving the default value into the getter.

Happy to go either way, not sure which one is better?